processor/schemaprocessor/internal/migrate/attributes.go

```diff
@@ -0,0 +1,131 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migrate // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/migrate"
+
+import (
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.uber.org/multierr"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/alias"
+)
+
+// AttributeKey is used to allow strings and type defined
+// strings to be used within the creation of creating an `AttributeChangeSet`
+type AttributeKey interface {
+	~string
+}
+
+// AttributeChangeSet represents an unscoped entry that can be applied.
+//
+// The listed changes are duplicated twice
+// to allow for simplified means of transition to or from a revision.
+type AttributeChangeSet struct {
+	updates  map[alias.AttributeKey]alias.AttributeKey
+	rollback map[alias.AttributeKey]alias.AttributeKey
+}
+
+// AttributeChangeSetSlice allows for `AttributeChangeSet`
+// to be chained together as they are defined within the schema
+// and be applied sequentially to ensure deterministic behavior.
+type AttributeChangeSetSlice []*AttributeChangeSet
+
+// NewAttributes allows for typed strings to be used as part
+// of the invocation that will be converted into the default string type.
+func NewAttributes[Key AttributeKey, Value AttributeKey](mappings map[Key]Value) *AttributeChangeSet {
+	attr := &AttributeChangeSet{
+		updates:  make(map[string]string, len(mappings)),
+		rollback: make(map[string]string, len(mappings)),
+	}
+	for k, v := range mappings {
+		attr.updates[string(k)] = string(v)
+		attr.rollback[string(v)] = string(k)
+	}
+	return attr
+}
+
+func (a *AttributeChangeSet) Apply(attrs pcommon.Map) error {
+	return a.do(StateSelectorApply, attrs)
+}
+
+func (a *AttributeChangeSet) Rollback(attrs pcommon.Map) error {
+	return a.do(StateSelectorRollback, attrs)
+}
+
+func (a *AttributeChangeSet) do(ss StateSelctor, attrs pcommon.Map) (errs error) {
+	var (
+		updated = make(map[string]struct{})
+		results = pcommon.NewMap()
+	)
+	attrs.Range(func(k string, v pcommon.Value) bool {
+		var (
+			key     string
+			matched bool
+		)
+		switch ss {
+		case StateSelectorApply:
+			key, matched = a.updates[k]
+		case StateSelectorRollback:
+			key, matched = a.rollback[k]
+		}
+		if matched {
+			k, updated[key] = key, struct{}{}
+		} else {
+			// TODO: Since the spec hasn't decided the behavior on what
```

Comment:
Link to an issue for deciding this in the comment?

Reply:
Not yet, was raised by @tigrannajaryan in a comment so I will raise it in the morning on the spec repo (or if anyone else gets to it before me).

Reply:
Added here https://github.com/open-telemetry/opentelemetry-specification/issues/3497

Reply:
I think this must be solved before marking the component as alpha and adding it to our official builds, but still I don't think it should block merging this PR in the interest of moving forward with other parts of the implementation